Filter configuration

sethfitz opened the pull request:
This implements a new component (`FilterConfigurator`) that allows the set of visible filters to be changed. The order in which they're displayed by the `Filter` component depends on the order in which they were enabled.

sethfitz committed:
Lock to node-6.x

sethfitz committed:
ESLint + config

sethfitz committed:
Implement filter configuration dialog

sethfitz committed:
Use formatMessage instead of FormattedMessage

sethfitz committed:
Move filter modal state to the URL

mojodna commented on src/components/discrete_filter.js:
```diff
@@ -42,7 +42,7 @@ class DiscreteFilter extends React.Component {
     fieldName: PropTypes.string.isRequired,
     checked: PropTypes.array,
-    values: PropTypes.array,
+    values: PropTypes.object,
```
This snuck in accidentally, but it quiets a warning I was seeing consistently.

mojodna commented on src/components/filter.js:
```diff
@@ -78,21 +79,24 @@ const Filter = ({
 }
 
-module.exports = pure(Filter)
+module.exports = pure(injectIntl(Filter))
```
This may conflict/complement `intlReducer`; I'm about to go read the docs about that.

mojodna replied:
`props.intl` is a function when pulling that from the Redux store.

sethfitz committed:
Unused / nonexistent

gmaclennan commented on src/components/filter.js:
```diff
@@ -78,21 +79,24 @@ const Filter = ({
-        leftIcon={<SettingsIcon style={style.listIcon} />}
-        primaryText={<FormattedMessage {...messages.changeFilters} />}
+      <FlatButton
+        label={intl.formatMessage(messages.changeFilters)}
```
Let's use `<FormattedMessage>` here to keep consistent with other code and avoid the extra code of injecting intl

mojodna replied:
Fixed in b9b092c at the cost of duplicating styles from material-ui.

gmaclennan commented on src/components/filter_configurator.js:
```diff
@@ -0,0 +1,102 @@
+    // FormattedMessage
+    return (
+      <Dialog
+        title={intl.formatMessage(messages.configureFilters)}
```
Just `<FormattedMessage>` renders a `<span>` because of the material-ui "bug" but we can just do `<h3><FormattedMessage {...messages.configureFilters} /></h3>` and style the h3 as we like, which I think is neater, and avoids the need to inject intl

gmaclennan replied:
Just a note, this isn't actually a bug, it's just the way material-ui works throughout. For label + title props, if you pass a string, it will wrap and style, but if you pass an element, then it just renders that and it's up to you to deal with styling. `<FormattedMessage>` returns a `<span>`, so the Dialog just renders that.

gmaclennan commented on src/components/filter_configurator.js:
```diff
@@ -0,0 +1,102 @@
+
+    // FormattedMessage
+    return (
+      <Dialog
```
We're doing a modal within a modal here. You can replace this with a `<Paper>` element, and remove open I think, and it will work because `<MatchModal>` takes care of the Modal stuff. See the code in the feature modal. You would just need to pass an onClose to the "Done" button.

mojodna replied:
Oh, that's what `MatchModal` is then? I didn't realize that it had UI.

gmaclennan replied:
yes it will render the child element as a modal, and handles animation etc. I'm not entirely happy with the way I implemented, but works for now. We can just swap out something similar if we change our modal implementation, and the rest of the code can stay the same

mojodna replied:
Fixed in 1a55ce3

gmaclennan commented on src/selectors/candidate_filters.js:
```diff
@@ -0,0 +1,13 @@
+
+const getFieldAnalysis = require('./field_analysis')
+
+const getCandidateFilters = createSelector(
```
The configurator isn't matching up with the state because of this selector, which is actually to do with some strange things that I do with the filter selector, which I am documenting and fixing, then this will work better for you.

mojodna replied:
This intentionally bypasses the synthetic “best fields” because if it includes them, it will allow them to be “removed” from the list of filters but won’t actually do anything since they’re not actually marked.

The UX would be to show 2 checked filters. You could check an additional one, but then the original 2 would disappear, showing only the new filter. If you tried to uncheck either of the 2 checked ones (if they’re the only ones available), nothing would happen.

mojodna replied:
Better after 82f558b

gmaclennan commented on src/components/filter.js:
```diff
@@ -78,21 +79,24 @@ const Filter = ({
-        innerDivStyle={style.listItemInner}
-        leftIcon={<SettingsIcon style={style.listIcon} />}
-        primaryText={<FormattedMessage {...messages.changeFilters} />}
+      <FlatButton
```
Let's match the style of the list item and icon that was there before. I'm pedantic about layout/spacing.

mojodna replied:
I am too. The `ListItem` wasn't firing events, presumably because it didn't have children, so I switched it to a button. I'll see what I can do to match the styling.

mojodna replied:
Fixed in b9b092c

sethfitz committed:
Removed random comment

sethfitz committed:
Change Filters styling

Uses a ListItem with a Link rather than a FlatButton. This simplifies
triggering opening the modal as well.

This also removes the intl injection, partially for consistently,
partially because it's no longer problematic.

mojodna commented on src/components/filter.js:
```diff
@@ -78,21 +79,24 @@ const Filter = ({
-        innerDivStyle={style.listItemInner}
-        leftIcon={<SettingsIcon style={style.listIcon} />}
-        primaryText={<FormattedMessage {...messages.changeFilters} />}
+      <FlatButton
```
Fixed in b9b092c

sethfitz committed:
Replace nested modal/dialog combo w/ a modal Card

sethfitz committed:
Match onCloseClick name

sethfitz committed:
Filter refactor

gmaclennan commented on src/components/filter.js:
```diff
@@ -81,16 +89,24 @@ const Filter = ({
-        primaryText={<FormattedMessage {...messages.changeFilters} />}
-      />
+      >
+        <Link
```
Needs a slight style tweak: the `<a>` rendered does not fill its parent, so if you click away from the text but within ListItem it does not fire.

We could tweak styling, or also do:

```
<Link to="/courses">{
  ({isActive, location, href, onClick, transition}) => 
    <ListItem
          innerDivStyle={style.listItemInner}
          leftIcon={<SettingsIcon style={style.listIcon} />}
          primaryText={<FormattedMessage {...messages.changeFilters} />}
          onClick={onClick}
        />
}</Link>
```
I don't completely understand the `<Link>` docs here, so probably would need a little test to see what the function args actually are https://react-router.now.sh/Link

mojodna replied:
`Link` creates a `<a>` behind the scenes with a real, live href (no JS). This means that no `onClick` handler needs to be passed in as props.

CSS tweaked in c37f181.

gmaclennan commented on src/selectors/filterable_fields.js:
```diff
@@ -0,0 +1,15 @@
+  getFieldAnalysis,
+  (fieldStats) => Object
+      .entries(fieldStats)
+      .filter(([k, v]) => v.filterType)
```
superfluous.

gmaclennan commented on src/reducers/visible_filters.js:
```diff
@@ -1,5 +1,11 @@
-const visibleFilters = (state = [], action) => {
-  return state
+const visibleFilters = (state = [], {type, payload}) => {
```
Having played with how this works, I think the default filters should only be set on initial load - as it is you can't get to the state of no visible filters in the UI, which is confusing I think.

I would change initial state here to `null` or `undefined` and then change https://github.com/mojodna/mapfilter/blob/82f558b8430701eb9cdaca9429fee1aae51e0636/src/selectors/visible_filters.js#L20-L24 to:

```
if (!visibleFilters) {
  visibleFilters = []
  if (dateFieldName) visibleFilters.push(dateFieldName)
  if (bestFilterFields[0]) visibleFilters.push(bestFilterFields[0].fieldname)
  return visibleFilters
}
```

mojodna replied:
Redux doesn't allow `undefined`, so it's explicitly set to null.

Done in a896290 and tested to get to the "no visible filters" state.

gmaclennan commented on src/components/filter_configurator.js:
```diff
@@ -0,0 +1,131 @@
+        <CardText>
+          <List>
+            {
+              // TODO allow these to be reordered
```
Not a priority for this trip, but I think when we do re-ordering we could also try it in the filter pane, rather than the settings page. That icon by each filter looks ripe for being dragged to re-order.

mojodna replied:
Totally. TODO moved to account for that.

gmaclennan commented on src/components/filter_configurator.js:
```diff
@@ -0,0 +1,131 @@
+                return (
+                  <ListItem
+                    key={field}
+                    primaryText={field}
```
Let's do:

```
  primaryText=<FormattedMessage {...msg('field_key')(row.key)} />
```

where: `const msg = require('../util/intl_helpers').createMessage`

Why? Well this custom `createMessage` creates a namespace for custom translations that can be a user-setting, and creates a default message by taking a guess at how to format things nicely. In the data sets we use, field keys and values are often computer names, rather than human-readable names. At some point we'll add a settings pane where you could set the custom translation e.g.

```
{
  en: {
    field_key: {
      happening: 'Activity',
      people: 'Monitor',
      today: 'Date of visit'
    }
}
```
And then things will look pretty. In the meantime it makes things title case and replaces `_` with ` `.

mojodna replied:
Nice. Done in d797664

gmaclennan reviewed (changes requested):
@mojodna this is really fantastic, I love how this works, I think our partners are really going to appreciate this. Just a few small tweaks and this is ready to merge.

gmaclennan committed:
Add title attibute to list item & internationalize strings

The CSS here truncates long values and adds an ellipsis. Adding a title attribute results in the browser showing a tooltip with the full text on hover.

gmaclennan commented:
One other tweak: I realize now that when field values are long that "ONLY" button becomes hard to see because it has a transparent and background and hover.

gmaclennan commented:
```
From 80081b39bf5babbdfb6c139cd9c4aa219a76f005 Mon Sep 17 00:00:00 2001
From: Gregor MacLennan <gmaclennan@digital-democracy.org>
Date: Sat, 10 Dec 2016 12:40:33 -0800
Subject: [PATCH] Add title attibute to list item & internationalize strings

The CSS here truncates long values and adds an ellipsis. Adding a title attribute results in the browser showing a tooltip with the full text on hover.
---
 src/components/discrete_filter.js | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/components/discrete_filter.js b/src/components/discrete_filter.js
index a896577..f4a1a4f 100644
--- a/src/components/discrete_filter.js
+++ b/src/components/discrete_filter.js
@@ -5,11 +5,12 @@ const makePure = require('recompose/pure').default
 const Checkbox = makePure(require('material-ui/Checkbox').default)
 const ListIcon = require('material-ui/svg-icons/action/list').default
 const {ListItem} = require('material-ui/List')
+const {FormattedMessage} = require('react-intl')
 const omit = require('lodash/omit')
 
 const ShowAllButton = require('./show_all_button')
 const OnlyButton = require('./only_button')
-const { t, titleCase } = require('../util/text_helpers')
+const msg = require('../util/intl_helpers').createMessage
 const { listStyles } = require('../styles')
 
 const styles = {
@@ -102,7 +103,7 @@ class DiscreteFilter extends React.Component {
     return (
       <ListItem
         innerDivStyle={listStyles.listItemInner}
-        primaryText={titleCase(fieldName)}
+        primaryText={<FormattedMessage {...msg('field_key')(fieldName)} />}
         leftIcon={<ListIcon style={listStyles.listIcon} />}
         initiallyOpen
         disabled
@@ -115,7 +116,8 @@ class DiscreteFilter extends React.Component {
             onMouseEnter={this.handleMouseEnter.bind(this, v)}
             onMouseLeave={this.handleMouseLeave}>
             <Checkbox
-              label={t(v)}
+              label={<FormattedMessage {...msg('field_value')(v)} />}
+              title={<FormattedMessage {...msg('field_value')(v)} />}
               value={v}
               style={styles.checkbox}
               iconStyle={styles.checkboxIcon}
-- 
2.8.1

```

gmaclennan commented:
Added a commit then realized I don't have commit access to your fork. See patch above ^^

sethfitz committed:
Ignore .env

mojodna commented:
Hmm:

> If checked, users with write access to digidem/mapfilter can add new commits to your filter-config branch. 

(It's checked.) This is a newish GitHub feature that I haven't used yet.

Patch applied locally anyway.

sethfitz committed:
Merge branch 'rewrite' into filter-config

sethfitz committed:
Superfluous

gmaclennan committed:
For example use un-transpiled code, remove browserify-css transform

sethfitz committed:
Make Link fill its parent

sethfitz committed:
Unused

sethfitz committed:
Move TODO

sethfitz committed:
Allow no visible filters to be set

sethfitz committed:
Future-proofing + more human-readable names

sethfitz committed:
Reduced 'only' button background opacity

mojodna commented:
"ONLY" button had its transparency dropped in 384f875. It's now only 10% transparent:

![image](https://cloud.githubusercontent.com/assets/45/21089104/18fb4050-bfea-11e6-9679-8b9c77e7a95d.png)

mojodna commented:
This is ready for another review.
